xugrid/regrid/structured.py

```diff
@@ -21,56 +28,619 @@ class StructuredGrid1d:
     bounds: (n, 2)
     """
 
-    def __init__(self, bounds):
+    def __init__(self, obj: Union[xr.DataArray, xr.Dataset], name: str):
+        bounds_name = f"{name}bounds"  # e.g. xbounds
+        size_name = f"d{name}"  # e.g. dx
+
+        index = obj.indexes[name]
+        # take care of potentially decreasing coordinate values
+        if index.is_monotonic_decreasing:
+            midpoints = index.values[::-1]
+            flipped = True
+            side = "right"
+        elif index.is_monotonic_increasing:
+            midpoints = index.values
+            flipped = False
+            side = "left"
+        else:
+            raise ValueError(f"{name} is not monotonic for array {obj.name}")
+
+        if bounds_name in obj.coords:
+            bounds = obj[bounds_name].values
+        else:
+            if size_name in obj.coords:
+                # works for scalar size and array size
+                size = np.abs(obj[size_name].values)
+            else:
+                # no bounds defined, no dx defined
+                # make an estimate of cell size
+                size = np.diff(midpoints)
+                # Check if equidistant
+                atolx = 1.0e-4 * size[0]
+                if not np.allclose(size, size[0], atolx):
+                    raise ValueError(
+                        f"DataArray has to be equidistant along {name}, or "
+                        f'explicit bounds must be given as "{name}bounds", or '
+                        f'cellsizes must be as "d{name}"'
+                    )
+
+            start = midpoints - 0.5 * size
+            end = midpoints + 0.5 * size
+            bounds = np.column_stack((start, end))
+
+        self.name = name
+        self.midpoints = midpoints
         self.bounds = bounds
+        self.flipped = flipped
+        self.side = side
+        self.grid = obj
```

Comment:
in here, if flipped= True then side =right. Is it always like that or can we have grids in which flipped is False and side =right ?
If they always correlate then it is not necessary to store them both, then you can only store flipped and make a function that returns right if flipped = false